// @igniteui/angular-templates/igx-ts/calendar/default/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const IgniteUIForAngularTemplate_1 = require("../../../IgniteUIForAngularTemplate");
class IgxCalendarTemplate extends IgniteUIForAngularTemplate_1.IgniteUIForAngularTemplate {
    constructor() {
        super(__dirname);
        this.userExtraConfiguration = {};
        this.components = ["Calendar"];
        this.controlGroup = "Scheduling";
        this.listInComponentTemplates = true;
        this.id = "calendar";
        this.projectType = "igx-ts";
        this.name = "Calendar";
        this.description = "IgxCalendar with single selection";
        this.dependencies = [
            { import: "IgxCalendarModule", from: "<%=igxPackage%>" }
        ];
        this.hasExtraConfiguration = true;
    }
    setExtraConfiguration(extraConfigKeys) {
        this.userExtraConfiguration = extraConfigKeys;
    }
    getExtraConfiguration() {
        return [{
                choices: ["Single selection", "Multi selection", "Range selection"],
                default: "Single selection",
                key: "selectionType",
                message: "Choose selection type for the igx-calendar",
                type: 0 /* Choice */
            }];
    }
    generateConfig(name, ...options) {
        let selectionType = "";
        let selectionMode = "";
        if (this.userExtraConfiguration["selectionType"]) {
            const type = this.userExtraConfiguration["selectionType"];
            selectionMode = type;
            switch (type) {
                case "Single selection":
                    selectionType = "";
                    break;
                case "Multi selection":
                    selectionType = 'selection="multi"';
                    break;
                case "Range selection":
                    selectionType = 'selection="range"';
                    break;
            }
        }
        const extraConfig = {
            selectionMode,
            selectionType
        };
        return super.generateConfig(name, { extraConfig });
    }
}
module.exports = new IgxCalendarTemplate();
