// @igniteui/angular-templates/igx-ts/input-group/default/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const IgniteUIForAngularTemplate_1 = require("../../../IgniteUIForAngularTemplate");
class IgxInputGroupTemplate extends IgniteUIForAngularTemplate_1.IgniteUIForAngularTemplate {
    constructor() {
        super(__dirname);
        this.components = ["Input Group"];
        this.controlGroup = "Data Entry & Display";
        this.listInComponentTemplates = true;
        this.id = "input-group";
        this.projectType = "igx-ts";
        this.name = "Input Group";
        this.description = "IgxInputGroup form view template";
        this.dependencies = [{
                import: [
                    "IgxButtonModule",
                    "IgxComboModule",
                    "IgxDatePickerModule",
                    "IgxIconModule",
                    "IgxInputGroupModule",
                    "IgxRippleModule",
                    "IgxTimePickerModule",
                    "IgxSelectModule"
                ],
                from: "<%=igxPackage%>"
            },
            { import: "FormsModule, ReactiveFormsModule", from: "@angular/forms" }
        ];
    }
}
module.exports = new IgxInputGroupTemplate();
