// @igniteui/angular-templates/igx-ts/grid/grid-batch-editing/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const IgniteUIForAngularTemplate_1 = require("../../../IgniteUIForAngularTemplate");
class IgxGridBatchEditingTemplate extends IgniteUIForAngularTemplate_1.IgniteUIForAngularTemplate {
    constructor() {
        super(__dirname);
        this.components = ["Grid"];
        this.controlGroup = "Grids & Lists";
        this.listInComponentTemplates = true;
        this.id = "grid-batch-editing";
        this.projectType = "igx-ts";
        this.name = "Grid Batch Editing";
        this.description = "Sample IgxGrid with batch editing";
        this.dependencies = [
            {
                import: [
                    "IgxGridModule",
                    "IgxFocusModule",
                    "IgxButtonModule",
                    "IgxDialogModule",
                    "IgxRippleModule"
                ],
                from: "<%=igxPackage%>"
            },
            { import: "FormsModule", from: "@angular/forms" }
        ];
    }
}
module.exports = new IgxGridBatchEditingTemplate();
