// @igniteui/cli-core/templates/BaseComponent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseComponent = void 0;
const path = require("path");
// cross-ref, don't use barrel exports
const Util_1 = require("../util/Util");
class BaseComponent {
    /**
     * Create a new Component
     * @param rootPath Root path for the Component directory (__dirname)
     */
    constructor(rootPath) {
        this.groupPriority = 0;
        this.basePath = __dirname;
        if (rootPath) {
            this.basePath = rootPath;
        }
        this.templates = this.loadTemplates();
    }
    get templateIds() {
        return this.templates.map(x => x.id);
    }
    loadTemplates() {
        let modulePath;
        const templates = [];
        const templatePaths = Util_1.Util.getDirectoryNames(this.basePath);
        for (const templatePath of templatePaths) {
            modulePath = path.join(this.basePath, templatePath);
            templates.push(require(modulePath));
        }
        return templates;
    }
}
exports.BaseComponent = BaseComponent;
