// @igniteui/angular-templates/igx-ts/map/default/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const IgniteUIForAngularTemplate_1 = require("../../../IgniteUIForAngularTemplate");
class IgxGeographicMapTemplate extends IgniteUIForAngularTemplate_1.IgniteUIForAngularTemplate {
    constructor() {
        super(__dirname);
        this.components = ["Geographic Map"];
        this.controlGroup = "Maps";
        this.listInComponentTemplates = true;
        this.id = "geographic-map";
        this.projectType = "igx-ts";
        this.name = "Geographic Map";
        this.description = "Basic IgxGeographicMap.";
        this.dependencies = [
            {
                import: ["IgxGeographicMapModule"],
                from: "igniteui-angular-maps"
            },
            {
                import: ["IgxDataChartInteractivityModule"],
                from: "igniteui-angular-charts"
            }
        ];
        this.packages = ["igniteui-angular-core@~13.1.0", "igniteui-angular-charts@~13.1.0", "igniteui-angular-maps@~13.1.0"];
    }
}
module.exports = new IgxGeographicMapTemplate();
