// igniteui-angular/migrations/common/ServerHost.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServerHost = void 0;
const pathFs = require("path");
const ts = require("typescript/lib/tsserverlibrary");
const tsUtils_1 = require("./tsUtils");
/**
 * Langauge server host is responsible for **most** of the FS operations / checks
 * Angular's Ivy LS sometimes bypasses these, calling path methods instead of tsLsHost operations
 */
class ServerHost {
    constructor(host) {
        this.host = host;
        this.args = ts.sys.args;
        this.newLine = ts.sys.newLine;
        this.useCaseSensitiveFileNames = ts.sys.useCaseSensitiveFileNames;
    }
    /**
     * Read a file's content from the Virtual Tree
     * If file does not exist in virtual tree, check in physical FS
     */
    readFile(path, encoding) {
        let content;
        // ensure the path is relative, so it can be found in the Tree, reflecting latest state
        path = pathFs.relative(this.getCurrentDirectory(), path);
        try {
            content = this.host.read(path).toString(encoding);
        }
        finally {
            // eslint-disable-next-line no-unsafe-finally
            return content || ts.sys.readFile(path, encoding);
        }
    }
    getFileSize(path) {
        return ts.sys.getFileSize(path);
    }
    watchFile(path, callback, pollingInterval) {
        return ts.sys.watchFile(path, callback, pollingInterval);
    }
    watchDirectory(path, callback, recursive) {
        return ts.sys.watchDirectory(path, callback, recursive);
    }
    resolvePath(path) {
        return ts.sys.resolvePath(path);
    }
    /**
     * Checks for file in Virtual Tree w/ relative path
     * If file does not exist in virtual tree, check in physical FS
     */
    fileExists(path) {
        // check for file in Tree, as schematics might need for check
        path = pathFs.relative(this.getCurrentDirectory(), path);
        let flag = false;
        try {
            // Tree.exists throws on invalid paths instead of returning false
            flag = this.host.exists(path);
        }
        finally {
            // eslint-disable-next-line no-unsafe-finally
            return flag || ts.sys.fileExists(path);
        }
    }
    directoryExists(path) {
        let exists;
        path = pathFs.relative(this.getCurrentDirectory(), path);
        try {
            exists = this.host.getDir(path) !== void 0;
        }
        finally {
            // eslint-disable-next-line no-unsafe-finally
            return exists || ts.sys.directoryExists(path);
        }
    }
    getExecutingFilePath() {
        return ts.sys.getExecutingFilePath();
    }
    getCurrentDirectory() {
        // both TS and NG lang serves work with absolute paths
        // we provide cwd instead of tree root so paths can be resolved to absolute ones
        return process.cwd();
    }
    /**
     * Get all subdirs of a directory from the Tree mapped to absolute paths
     */
    getDirectories(path) {
        // check directory contents in Tree (w/ relative paths)
        path = pathFs.relative(this.getCurrentDirectory(), path);
        // return directory contents w/ absolute paths for LS
        return this.host.getDir(path).subdirs.map(e => pathFs.resolve(e));
    }
    /**
     * Get all files of a directory from the Tree mapped to absolute paths
     */
    readDirectory(path) {
        // check directory contents in Tree (w/ relative paths)
        path = pathFs.relative(this.getCurrentDirectory(), path);
        // return directory contents w/ absolute paths for LS
        return this.host.getDir(path).subfiles.map(e => pathFs.resolve(e));
    }
    require(initialPath, moduleName) {
        try {
            const paths = [initialPath];
            if (moduleName === tsUtils_1.CUSTOM_TS_PLUGIN_NAME) {
                moduleName = tsUtils_1.CUSTOM_TS_PLUGIN_PATH;
                paths.push(__dirname);
            }
            const modulePath = require.resolve(moduleName, { paths });
            return {
                module: require(modulePath),
                error: undefined,
            };
        }
        catch (e) {
            return {
                module: undefined,
                error: e,
            };
        }
    }
    getModifiedTime(path) {
        return ts.sys.getModifiedTime(path);
    }
    realpath(path) {
        return ts.sys.realpath(path);
    }
    createSHA256Hash(data) {
        return ts.sys.createSHA256Hash(data);
    }
    //#region Not implemented methods
    write(_data) {
        throw new Error('Method "write" not implemented.');
        // ts.sys.write(data);
    }
    writeOutputIsTTY() {
        throw new Error('Method "writeOutputIsTTY" not implemented.');
        // return ts.sys.writeOutputIsTTY();
    }
    writeFile(_path, _data, _writeByteOrderMark) {
        throw new Error('Method "writeFile" not implemented.');
        // return ts.sys.writeFile(path, data, writeByteOrderMark);
    }
    createDirectory(_path) {
        throw new Error('Method "createDirectory" not implemented.');
        // return ts.sys.createDirectory(path);
    }
    setModifiedTime(_path, _time) {
        throw new Error('Method "setModifiedTime" not implemented.');
        // return ts.sys.setModifiedTime(path, time);
    }
    deleteFile(_path) {
        throw new Error('Method "deleteFile" not implemented.');
        // return ts.sys.deleteFile(path);
    }
    createHash(_data) {
        throw new Error('Method "createHash" not implemented.');
        // return ts.sys.createHash(data);
    }
    getMemoryUsage() {
        throw new Error('Method "getMemoryUsage" not implemented.');
        // return ts.sys.getMemoryUsage();
    }
    exit(_exitCode) {
        throw new Error('Method "exit" not implemented.');
        // return ts.sys.exit(exitCode);
    }
    setTimeout(_callback, _ms, ..._args) {
        throw new Error('Method "setTimeout" not implemented.');
        // return ts.sys.setTimeout(callback, ms, ...args);
    }
    clearTimeout(_timeoutId) {
        throw new Error('Method "clearTimeout" not implemented.');
        // return ts.sys.clearTimeout(timeoutId);
    }
    clearScreen() {
        throw new Error('Method "clearScreen" not implemented.');
        // return ts.sys.clearScreen();
    }
    base64decode(_input) {
        throw new Error('Method "base64decode" not implemented.');
        // return ts.sys.base64decode(input);
    }
    base64encode(_input) {
        throw new Error('Method "base64encode" not implemented.');
        // return ts.sys.base64encode(input);
    }
    setImmediate(_callback, ..._args) {
        throw new Error('Method "setImmediate" not implemented.');
        // return setImmediate(callback, ...args);
    }
    clearImmediate(_timeoutId) {
        throw new Error('Method "clearImmediate" not implemented.');
        // return clearImmediate(timeoutId);
    }
}
exports.ServerHost = ServerHost;
