// @igniteui/angular-templates/igx-ts/dock-manager/default/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const IgniteUIForAngularTemplate_1 = require("../../../IgniteUIForAngularTemplate");
const package_resolve_1 = require("../../../package-resolve");
class IgcDockManagerTemplate extends IgniteUIForAngularTemplate_1.IgniteUIForAngularTemplate {
    constructor() {
        super(__dirname);
        this.components = ["Dock Manager"];
        this.controlGroup = "Layouts";
        this.listInComponentTemplates = true;
        this.id = "dock-manager";
        this.projectType = "igx-ts";
        this.name = "Dock Manager";
        this.description = "basic IgcDockManager";
        this.dependencies = [
            { import: "<%=ClassName%>Module", from: "./src/app/<%=path%>/<%=filePrefix%>.module.ts" }
        ];
        // "igniteui-dockmanager@~1.0.0":
        this.packages = [`${(0, package_resolve_1.resolveIgxPackage)(package_resolve_1.NPM_DOCK_MANAGER)}@~1.6.0`];
    }
    addClassDeclaration(mainModule, projPath, name, modulePath) {
        // not applicable with custom module
    }
}
module.exports = new IgcDockManagerTemplate();
